Explicitly set which DQM modules use LuminosityBlock transitions

Dr15Jones opened the pull request:
This change identifies which DQM modules are using LuminosityBlock transitions and explicitly changes the base class they use. The using modules were classified into two categories:

1. Modules creating Lumi based histograms (ie. they call `setLumiFlag`) and
1. Modules which do other work at Lumi boundaries

Separating the two makes it possible to only have those modules which actually have Lumi based histograms to be LuminosityBlockProducers. This decreases the number of data products in the `edm::LuminosityBlock` and allows for more efficient running of the transition.

Dr15Jones committed:
CTPPS DQM modules now use LuminosityBlockCache

Using the LuminosityBlockCache allows the framework to properly
utilize concurrent LuminosityBlocks in a job.

Dr15Jones committed:
Removed printout of LuminosityBlock count from HiggsDQM

The code was requiring the framework to synchronize on Lumi
boundaries. The printout was not going to be seen anyway in
standard jobs.

Dr15Jones committed:
QcdLowPtDQM now supports concurrent LuminosityBlocks

Switching to using a LuminosityBlockCache allows the framework to
efficiently process concurrent LuminosityBlocks.

Dr15Jones committed:
DQMEDAnalyzer no longer uses LuminosityBlock transitions

DQM modules needed LuminosityBlock transitions must now explicitly
inherit from one::DQMEDAnalyzer<>.

Dr15Jones committed:
Reorganized modules in DQMServices/Components

Moved all non-externally linked modules to plugins directory.

Dr15Jones committed:
Module cleanup in DQM/L1TMonitor

Made the following modifications in L1TRCT and L1TRPCTF
- removed unused member variables
- put globals into anonymous namespace to avoid poluting the global
  namespace
- remove trivial member functions associated with Run and LuminosityBlock
  transitions.

Dr15Jones committed:
Made DQMFileSaver::fillJson a stand alone function

Since fillJson is used by DQMFileSaver and JsonWritingTimeoutPoolOutputModule
it is better to make it a stand alone function.

Dr15Jones committed:
DQMEDAnalyzers now explicitly say if they use a LuminosityBlock

All DQMEDAnalyzers which rely on seeing the LuminosityBlock
transition explictly tell that to the framework. In addition,
the modules which use setLumiFlag on a MonitorElement are also
explicitly declared. Only those modules will have a DQMToken
put into the LuminosityBlock during the end transition.

These changes are needed to transition to using multiple concurrent
LuminosityBlocks within a job.

Dr15Jones committed:
Fixed finding header files from DQMServices/Components

Modules in DQMServices/Components were moved to the plugin directory.
Some includes were missed during the migration.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-25188/7183

Code check has found code style and quality issues which could be resolved by applying a patch in https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-25188/7183/git-diff.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-25188/7183/git-diff.patch | patch -p1`

You can run `scram build code-checks` to apply code checks directly

Dr15Jones committed:
Fixes from code-checks

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-25188/7184

cmsbuild commented:
A new Pull Request was created by  @Dr15Jones (Chris Jones) for master.

It involves the following packages:

CalibTracker/SiPixelQuality
Calibration/LumiAlCaRecoProducers
DQM/BeamMonitor
DQM/CTPPS
DQM/DTMonitorModule
DQM/EcalMonitorTasks
DQM/EcalPreshowerMonitorModule
DQM/HcalCommon
DQM/L1TMonitor
DQM/Physics
DQM/PixelLumi
DQM/RPCMonitorClient
DQM/RPCMonitorDigi
DQM/SiPixelMonitorDigi
DQM/SiPixelMonitorRawData
DQM/SiStripMonitorDigi
DQM/SiStripMonitorHardware
DQM/TrackingMonitor
DQM/TrigXMonitor
DQMOffline/EGamma
DQMOffline/L1Trigger
DQMServices/Components
DQMServices/Core
DQMServices/StreamerIO


@cmsbuild, @andrius-k, @kmaeshima, @schneiml, @tocheng, @nsmith-, @rekovic, @franzoni, @jfernan2, @thomreis, @pohsun, @lpernie can you please review it and eventually sign? Thanks.
@rappoccio, @forthommel, @argiro, @fioriNTU, @thomreis, @threus, @acimmino, @kreczko, @hdelanno, @makortel, @rociovilar, @barvic, @DryRun, @rovere, @tocheng, @deguio, @jandrea, @idebruyn, @mmusich, @mtosi, @jan-kaspar, @dkotlins this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: 9324b2ae5331d847e5d2d6b98a10cf38880ce1f9

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25188/31571/summary.html

I found follow errors while testing this PR

**Failed tests:  ClangBuild**

* **Clang**:

I found compilation warning while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 16 COMPILER='llvm compile'

```
 See details on the summary page.

cmsbuild commented:
Comparison not run due to Build errors/Fireworks only changes/No short matrix requested (RelVals and Igprof tests were also skipped)

Dr15Jones committed:
Do not hide virtual function openFile

Changing the function name to openFile_ avoids the hiding. Not able
to do 'using openFile' because the function is private in the base
class.

Dr15Jones committed:
Mark functions as override

Dr15Jones committed:
Removed 'using namespace' from headers

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-25188/7185

cmsbuild commented:
Pull request #25188 was updated. @cmsbuild, @andrius-k, @kmaeshima, @schneiml, @tocheng, @nsmith-, @rekovic, @franzoni, @jfernan2, @thomreis, @pohsun, @lpernie can you please check and sign again.

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 248acb6beb8cc8b128ab71be1d83f72f7462b5c5
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25188/31573/summary.html

cmsbuild commented:
Comparison job queued.

schneiml commented:
Thanks Chris, this is a cleanup I wanted to have for a while. Were some of the changes done by machine or are all the commits manual work? I am a bit scared by the non-trivial changes, which are always risky when done by hand... would be easier to review if we can separate out the mechanical changes.

As mentioned before, I am not that happy with making things explicitly `one::`, since I don't want them to be `one::` anymore in the mid-term future. Though changing back things should now be a matter of find&replace and not a big deal. 

Also interesting is the final number of modules of the two types. If there are not many, maybe we can remove some features from the DQM infrastructure (and find different solutions). Though, also in the mid-term future, I want *all* DQM output to be produced on a finer-than-runs granularity, which *will* mean throwing a large number of products into the Lumi blocks. If there is any constraint on what can be done using Products (how many different ones to have etc.) we should talk about that now, since my wish was to replace all of `DQMStore` with `edm`-based products and explicit dependencies.

Dr15Jones commented:
@schneiml all changes were done by 'hand', although I leveraged the compiler and `git grep` where ever I could. The way I proceeded was
1. changed DQMEDAnalyzer to be `one::DQMEDAnalyzer<>` (i.e. dropping `one::DQMLuminosityBlockElements`). This immediately caused all DQM modules which declared the use of `LuminosityBlock` callbacks with `override` to fail to compile.
1. used `git grep setLumiFlag` to find all modules which are generated per Lumi MonitorElements. For all of those modules, I made them inherit explicitly from `one::DQMEDAnalyzer<one::DQMLuminosityBlockElements>`. 
1. systematically went through the list of failing to compile modules to see which ones actually needed LuminosityBlock transitions. Based on what I found I
   1. removed unnecessary calls
   1. switched to using a LuminosityBlock cache if such a conversion was 'easy'
   1. made it a `one::WatchLuminosityBlocks` to return to the behavior it had before the change

So it is possible I missed DQMEDAnalyzers which used LuminosityBlock transitions but did not use the `override` keyword.

Dr15Jones commented:
> Though, also in the mid-term future, I want all DQM output to be produced on a finer-than-runs granularity, which will mean throwing a large number of products into the Lumi blocks. If there is any constraint on what can be done using Products (how many different ones to have etc.) we should talk about that now, since my wish was to replace all of DQMStore with edm-based products and explicit dependencies.

My take is we should not request the framework to do unnecessary extra work or use extra resources. So if a module does not actually produce something on a LuminosityBlock boundary, we shouldn't require the framework to reserve memory for that product nor call the module on that transition. Right now the DQM system is asking the framework to do a bunch of unnecessary work. In addition, the way it is requesting the unnecessary work is also keeping the framework from being able to efficiently process concurrent LuminosityBlocks.

Dr15Jones commented:
@schneiml  The goal I want to see is before the end of 2019 I want the prompt RECO job we will run at the Tier 0 for Run 3 to be able to efficiently use concurrent LuminosityBlocks in order to improve our CPU efficiency and increase our event throughput.  That time scale would give us plenty of time to stress test that new configuration before data taking. What is preventing that from being true with our current code is the DQM modules.

The work I was doing here was meant as a stepping stone towards that goal which could be done on that time scale (actually, could be done by the beginning of 2019). One possibility is we use what I did here as an intermediate step towards the larger changes you envision. If we can get things to work using `one` modules then the changes you want to make would not have to be finished before 2019 since we'd already have a working solution we can use before the larger change.

Dr15Jones commented:
@schneiml how do we move forward?

schneiml commented:
Nothing immediatly wrong with this PR, I just wanted you to know about my plans. For now, we definitely want the changes in here (even if we might undo some of it at some point, that remains to be seen).

Dr15Jones commented:
@schneiml Is there anything you need done before you sign this pull request?

andrius-k commented:
+1

Dr15Jones commented:
Please review @franzoni @pohsun @tocheng @lpernie @rekovic @thomreis @nsmith-

Dr15Jones commented:
please review @franzoni @pohsun @tocheng @lpernie @rekovic @thomreis @nsmith-

thomreis commented on DQM/L1TMonitor/src/L1TRCT.cc:
```diff
@@ -47,20 +48,6 @@ L1TRCT::L1TRCT(const ParameterSet & ps) :
     std::cout << "L1TRCT: constructor...." << std::endl;
-
-  outputFile_ =
-      ps.getUntrackedParameter < std::string > ("outputFile", "");
```
If this parameter is not used anymore it should also be removed from the config files.

thomreis commented on DQM/L1TMonitor/src/L1TRCT.cc:
```diff
@@ -47,20 +48,6 @@ L1TRCT::L1TRCT(const ParameterSet & ps) :
-  }
-
-  bool disable =
-      ps.getUntrackedParameter < bool > ("disableROOToutput", false);
```
If this parameter is not used anymore it should also be removed from the config files.

thomreis commented on DQM/L1TMonitor/src/L1TRPCTF.cc:
```diff
@@ -32,15 +32,6 @@ L1TRPCTF::L1TRPCTF(const ParameterSet& ps)
-    cout << "L1T Monitoring histograms will be saved to " << outputFile_.c_str() << endl;
-  }
-
-  bool disable = ps.getUntrackedParameter<bool>("disableROOToutput", false);
```
If this parameter is not used anymore it should also be removed from the config files.

thomreis commented on DQM/L1TMonitor/src/L1TRPCTF.cc:
```diff
@@ -32,15 +32,6 @@ L1TRPCTF::L1TRPCTF(const ParameterSet& ps)
 
   if(verbose_) cout << "L1TRPCTF: constructor...." << endl;
 
-  outputFile_ = ps.getUntrackedParameter<string>("outputFile", "");
```
If this parameter is not used anymore it should also be removed from the config files.

thomreis commented on DQM/L1TMonitor/src/L1TdeRCT.cc:
```diff
@@ -122,20 +123,6 @@ L1TdeRCT::L1TdeRCT(const ParameterSet & ps) :
     std::cout << "L1TdeRCT: constructor...." << std::endl;
 
-  outputFile_ =
-      ps.getUntrackedParameter < std::string > ("outputFile", "");
```
If this parameter is not used anymore it should also be removed from the config files.

thomreis commented on DQM/L1TMonitor/src/L1TdeRCT.cc:
```diff
@@ -122,20 +123,6 @@ L1TdeRCT::L1TdeRCT(const ParameterSet & ps) :
-  }
-
-  bool disable =
-      ps.getUntrackedParameter < bool > ("disableROOToutput", false);
```
If this parameter is not used anymore it should also be removed from the config files.

thomreis commented on DQMOffline/L1Trigger/src/L1TSync_Offline.cc:
```diff
@@ -202,14 +202,6 @@ L1TSync_Offline::L1TSync_Offline(const ParameterSet & pset) :
-    std::cout << "L1T Monitoring histograms will be saved to " <<	m_outputFile.c_str() << std::endl;
-  }
-
-  bool disable = pset.getUntrackedParameter < bool > ("disableROOToutput", false);
```
If this parameter is not used anymore it should also be removed from the config files.

thomreis commented on DQMOffline/L1Trigger/src/L1TSync_Offline.cc:
```diff
@@ -202,14 +202,6 @@ L1TSync_Offline::L1TSync_Offline(const ParameterSet & pset) :
 
   }
 
-  m_outputFile = pset.getUntrackedParameter < std::string > ("outputFile","");
```
If this parameter is not used anymore it should also be removed from the config files.

Dr15Jones commented:
@thomreis could we do the removal from the config in another pull request? This one has already been waiting for more than a month and is important to get concurrent lumis usable.

Dr15Jones commented:
@franzoni @lpernie @pohsun @tocheng @fabiocos please review this waiting pull request for ALCA.

lpernie commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

Dr15Jones commented:
@fabiocos what is your plan for integrating this?

fabiocos commented:
+1

I understand that this update is ok for DQM so far, and @thomreis  L1T cleaning may be done afterwards for simplicity
